### PR TITLE
fix: remove panic from read_last_checkpoint

### PR DESCRIPTION
**Pull request by zachschuermann**
## What changes are proposed in this pull request?
this PR does two main things:
1. remove the `expect` which panics in `read_last_checkpoint`
2. minor refactor of the method in general. The API was simplified: instead of returning `DeltaResult<Option<LastCheckpointHint>>` we now just return `DeltaResult<LastCheckpointHint>` and the consumer should just warn on `Err` (snapshot code updated appropriately). This more clearly communicates that it is indeed a fallible operation but we include docs saying that callers should just not propagate the error up and instead probably just ignore/warn (which we do in snapshot code that consumes it)

a couple other minor changes were included
1. derive `PartialEq, Eq` on `LastCheckpointHint`
2. generally improve testing

## How was this change tested?
modified UT

### CODE REVIEW

File: kernel/src/snapshot.rs [hunk 1]
```diff
@@ -387,29 +393,22 @@ pub(crate) struct LastCheckpointHint {
...
-        Ok(data) => Ok(serde_json::from_slice(&data)
-            .inspect_err(|e| warn!("invalid _last_checkpoint JSON: {e}"))
-            .ok()),
-        Err(Error::FileNotFound(_)) => Ok(None),
-        Err(err) => Err(err),
-    }
+    let last_checkpoint_data = storage
+        .read_files(vec![(file_path, None)])?
+        .next()
+        .ok_or_else(|| Error::generic("_last_checkpoint file is empty"))?;
```
Comment (nicklan) [inline kernel/src/snapshot.rs]: I don't think it's that the file is necessarily empty right? We should propagate the error. Assuming we have an into just:
```suggestion
        .read_files(vec![(file_path, None)])?
        .next()?;
```

Comment (zachschuermann) [reply kernel/src/snapshot.rs]: Ah this is to convert the `Option` to an error in case read_files returns an empty iterator

Comment (nicklan) [reply kernel/src/snapshot.rs]: Ohh hah, right, got it

**Review by nicklan — APPROVED** [review]

File: kernel/src/snapshot.rs [hunk 2]
```diff
@@ -393,22 +393,23 @@ pub(crate) struct LastCheckpointHint {
 /// the read. Thus, the semantics of this function are to return `None` if the file is not found or
 /// is invalid JSON. Unexpected/unrecoverable errors are returned as `Err` case and are assumed to
 /// cause failure.
-///
-/// TODO: java kernel retries three times before failing, should we do the same?
+// TODO: weird that we propagate FileNotFound as part of the iterator instead of top-level
+// result coming from storage.read_files
```
Comment (OussamaSaoudi) [inline kernel/src/snapshot.rs]: Could you create an issue so this is tracked?

File: kernel/src/snapshot.rs [hunk 3]
```diff
@@ -366,7 +366,7 @@ impl Snapshot {
 }
 
 // Note: Schema can not be derived because the checkpoint schema is only known at runtime.
-#[derive(Debug, Deserialize, Serialize)]
+#[derive(Debug, Deserialize, Serialize, PartialEq, Eq)]
```
Comment (OussamaSaoudi) [inline kernel/src/snapshot.rs:374]: qq: What does it mean for two `Schema`s to be equal? Is metadata included in this? 

In particular, what kind of behaviour do we want here, and should we protect against future changes to the definition of Eq for Schema?

File: kernel/src/snapshot.rs [hunk 4]
```diff
@@ -806,19 +808,19 @@ mod tests {
...
     }
 
     fn valid_last_checkpoint() -> Vec<u8> {
-        r#"{"size":8,"size_in_bytes":21857,"version":1}"#.as_bytes().to_vec()
+        r#"{"size":8,"sizeInBytes":21857,"version":1}"#.as_bytes().to_vec()
     }
 
     #[test]
-    fn test_read_table_with_invalid_last_checkpoint() {
+    fn test_read_table_with_last_checkpoint() {
```
Comment (OussamaSaoudi) [inline kernel/src/snapshot.rs:849]: For completeness, can we also add a test case for empty _last_checkpoint?

File: kernel/src/snapshot.rs (hunk 2 above)
Comment (zachschuermann) [reply kernel/src/snapshot.rs]: #1047

File: kernel/src/snapshot.rs (hunk 3 above)
Comment (zachschuermann) [reply kernel/src/snapshot.rs:374]: > qq: What does it mean for two Schemas to be equal? Is metadata included in this?

good question - it's derived all the way down (`StructType`, `StructField`, `DataType`, etc. all just derive `PartialEq`/`Eq`) so yes, the metadata in struct fields must be identical to be equal

> In particular, what kind of behaviour do we want here, and should we protect against future changes to the definition of Eq for Schema?

hm I might need help understanding this - is this asking how we can prevent changing our definition (and therefore implementation) of `(Partial)Eq` in the future?

Comment (OussamaSaoudi) [reply kernel/src/snapshot.rs:374]: >  is this asking how we can prevent changing our definition (and therefore implementation) of (Partial)Eq in the future

Yep correct. sry for being imprecise. the concern is that schema equality could evolve once we start stripping delta metadata from the schema.

ig for now we can go with this, but might be something to revisit.

**Review by OussamaSaoudi — APPROVED** [review]
